### PR TITLE
Improved granularity of error codes closes #2847

### LINKED ISSUE #2847: Return more accurate error codes from enproxy server
We're currently returning both 403 and 502 for a bunch of different error cases, many of which are not related to those error cases.

### PR DESCRIPTION
**Pull request by myleshorton**

### CODE REVIEW

**Comment by myleshorton** [comment]
Hey I'm assigning this one over to you @xiam whenever you have a sec!

File: src/github.com/getlantern/enproxy/proxy.go [hunk 1]
```diff
@@ -311,25 +311,31 @@ func (p *Proxy) handleRead(resp http.ResponseWriter, req *http.Request, lc *lazy
...
+}
+
+// newOutgoingConn creates a new outoing connection and stores it in the connection cache.
+func (p *Proxy) newOutgoingConn(id string, addr string, req *http.Request, resp http.ResponseWriter) (l *lazyConn, isNew bool, err error) {
+	if p.Allow != nil {
+		log.Trace("Checking if connection is allowed")
+		code, err := p.Allow(req, addr)
+		if err != nil {
+			respond(code, resp, err.Error())
+			return nil, false, fmt.Errorf("Not allowed: %v", err)
```
Comment (myleshorton) [inline src/github.com/getlantern/enproxy/proxy.go]: This refactoring into a subroutine is not strictly necessary...

File: src/github.com/getlantern/enproxy/proxy.go [hunk 2]
```diff
@@ -347,14 +353,6 @@ func clientIpFor(req *http.Request) string {
...
 }
 
-func badGateway(resp http.ResponseWriter, msg string) {
-	respond(http.StatusBadGateway, resp, fmt.Sprintf("Enproxy server responding Bad Gateway: %s", msg))
-}
-
-func forbidden(resp http.ResponseWriter, msg string) {
-	respond(http.StatusForbidden, resp, fmt.Sprintf("Enproxy server responding Forbidden: %s", msg))
-}
-
```
Comment (myleshorton) [inline src/github.com/getlantern/enproxy/proxy.go]: Getting rid of these convenience functions forces the caller to think a bit more about the response code and error, which I think is good.

File: src/github.com/getlantern/flashlight/server/server.go [hunk 3]
```diff
@@ -139,22 +139,22 @@ func (server *Server) ListenAndServe(updateConfig func(func(*ServerConfig) error
...
-					return err
+					return http.StatusForbidden, fmt.Errorf("Port not allowed in %v", destAddr)
 				}
 			}
 
 			if server.AllowedCountries != nil {
 				err := server.checkForDisallowedCountry(req)
 				if err != nil {
-					return err
+					return http.StatusForbidden, fmt.Errorf("Origin country not allowed")
```
Comment (myleshorton) [inline src/github.com/getlantern/flashlight/server/server.go]: These are the only two that really should be considered 403 forbidden, but now the message body in the response will tell us much more.

File: src/github.com/getlantern/fronted/server.go [hunk 4]
```diff
@@ -147,17 +147,17 @@ func (server *Server) Serve(l net.Listener) error {
...
-				return err
+				return code, err
 			}
 
 			if server.Allow != nil {
 				return server.Allow(req, destAddr)
 			}
 
-			return nil
+			return http.StatusOK, nil
```
Comment (myleshorton) [inline src/github.com/getlantern/fronted/server.go]: The OK status isn't really used in practice, but we need to return something.

File: src/github.com/getlantern/fronted/server.go [hunk 5]
```diff
@@ -180,22 +180,24 @@ func (server *Server) dialDestination(addr string) (net.Conn, error) {
...
-		return fmt.Errorf("Unable to split host and port for %v: %v", addr, err)
+		return http.StatusBadRequest, fmt.Errorf("Enproxy unable to split host and port for %v: %v", addr, err)
 	}
 
 	ipAddr, err := net.ResolveIPAddr("ip", host)
 	if err != nil || len(ipAddr.IP) == 0 {
-		return fmt.Errorf("Unable to resolve destination IP addr %v: %v", host, err)
+		// It's not really clear what we should return if we can't resolve the DNS, so
+		// might as well go with status tea pot!
+		return http.StatusTeapot, fmt.Errorf("Unable to resolve destination IP addr %v: %v", host, err)
```
Comment (myleshorton) [inline src/github.com/getlantern/fronted/server.go]: Teapot!

**Comment by xiam** [comment]
Looks good @myleshorton, merged!

**Pull request merged**